java/ql/src/experimental/Security/CWE/CWE-321/HardcodedJwtKey.qll

```diff
@@ -0,0 +1,156 @@
+  VerifyTokenMethod() {
+    this.getDeclaringType() instanceof JwtVerifier and
+    this.hasName("verify")
+  }
+}
+
+/**
+ * A data flow source for JWT token signing vulnerabilities.
+ */
+abstract class JwtKeySource extends DataFlow::Node { }
+
+/**
+ * A data flow sink for JWT token signing vulnerabilities.
+ */
+abstract class JwtTokenSink extends DataFlow::Node { }
+
+private predicate isTestCode(Expr e) {
+  e.getFile().getAbsolutePath().toLowerCase().matches("%test%") and
+  not e.getFile().getAbsolutePath().toLowerCase().matches("%ql/test%")
+}
+
+/**
+ * A hardcoded string literal as a source for JWT token signing vulnerabilities.
+ */
+class HardcodedKeyStringSource extends JwtKeySource {
+  HardcodedKeyStringSource() {
+    this.asExpr() instanceof CompileTimeConstantExpr and
+    not isTestCode(this.asExpr())
+  }
+}
+
+/**
+ * An expression used to sign JWT tokens as a sink of JWT token signing vulnerabilities.
+ */
+private class SignTokenSink extends JwtTokenSink {
+  SignTokenSink() {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof SignTokenMethod and
+      this.asExpr() = ma.getArgument(0)
+    )
+  }
+}
+
+/**
+ * An expression used to verify JWT tokens as a sink of JWT token signing vulnerabilities.
+ */
+private class VerifyTokenSink extends JwtTokenSink {
+  VerifyTokenSink() {
+    exists(MethodAccess ma |
+      ma.getMethod() instanceof VerifyTokenMethod and
+      this.asExpr() = ma.getQualifier()
+    )
+  }
+}
+
+/**
+ * A configuration depicting taint flow for checking JWT token signing vulnerabilities.
+ */
+class HardcodedJwtKeyConfiguration extends TaintTracking::Configuration {
+  HardcodedJwtKeyConfiguration() { this = "Hard-coded JWT Signing Key" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof JwtKeySource }
+
+  override predicate isSink(DataFlow::Node sink) { sink instanceof JwtTokenSink }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node prev, DataFlow::Node succ) {
+    exists(MethodAccess ma |
+      (
+        ma.getMethod() instanceof GetSecretMethod or
+        ma.getMethod() instanceof RequireMethod
+      ) and
+      prev.asExpr() = ma.getArgument(0) and
+      succ.asExpr() = ma
+    )
+  }
+}
+
+/** Taint model related to verifying JWT tokens. */
+private class VerificationFlowStep extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "com.auth0.jwt.interfaces;Verification;true;build;;;Argument[-1];ReturnValue;taint",
+        "com.auth0.jwt.interfaces;Verification;true;" +
+          ["acceptLeeway", "acceptExpiresAt", "acceptNotBefore", "acceptIssuedAt", "ignoreIssuedAt"]
+          + ";;;Argument[-1];ReturnValue;taint",
+        "com.auth0.jwt.interfaces;Verification;true;with" +
+          [
+            "Issuer", "Subject", "Audience", "AnyOfAudience", "ClaimPresence", "Claim",
+            "ArrayClaim", "JWTId"
+          ] + ";;;Argument[-1];ReturnValue;taint"
```

Comment:
I created a test case for this scenario, which is:
https://github.com/github/codeql/blob/75e7148912e3e70f322cced85d49319adc6f4a60/java/ql/test/experimental/query-tests/security/CWE-321/HardcodedJwtKey.java#L42-L44

The method call `JWT.require(Algorithm algorithm)` returns an object of the type `Verification`. And frequently the object is initialized in method chaining as shown in the test case. Without the `taint`/`value` steps, this test case cannot be detected. 

Does this answer your question? Or is there a better way to model this?

Thanks,
@luchua-bc

Reply:
Yes, this answer the question: the `Verification` object is tainted by the argument of the `JWT.require` call. I didn't spot it at first because it's a query-specific additional taint step, whereas the rest of `Verification`-related steps are defined as CSV rows.

Thanks for the clarification :-)